### PR TITLE
HDDS-5869. Added support for stream on S3Gateway write path

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

Added support for stream on S3Gateway write path

This section does not include multipart uploads.

ZERO copy is not used, and subsequent optimization is performed.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5869

## How was this patch tested?

Existing test
ci: https://github.com/guohao-rosicky/ozone/actions/runs/5364940200

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
hi @szetszwo PTAL thanks.

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpointStreaming.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        byte[] subBuffer = new byte[nn];
+        System.arraycopy(buffer, 0, subBuffer, 0, nn);
+        writeByteBuffer = ByteBuffer.wrap(subBuffer, 0, nn);
+      } else {
+        writeByteBuffer = ByteBuffer.wrap(buffer, 0, nn);
+      }
+      streamOutput.write(writeByteBuffer, 0, nn);
+      total += nn;
+    } while (total != length);
+    return total;
```
Comment (szetszwo) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpointStreaming.java:120]: The method can be simplified to 
```java
    final byte[] buffer = new byte[bufferSize];
    long n = 0;
    while (n < length) {
      final int toRead = Math.toIntExact(Math.min(bufferSize, length - n));
      final int readLength = body.read(buffer, 0, toRead);
      if (readLength == -1) {
        break;
      }
      streamOutput.write(ByteBuffer.wrap(buffer, 0, readLength));
      n += readLength;
    }
    return n;
```

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java [hunk 2]
```diff
@@ -233,11 +255,19 @@ public Response put(
           .equals(headers.getHeaderString("x-amz-content-sha256"))) {
         body = new SignedChunksInputStream(body);
       }
+      long putLength = 0;
+      if (datastreamEnabled && !enableEC) {
```
Comment (szetszwo) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java:259]: How about we use Streaming only if it has more than one chunk?
```java
      if (datastreamEnabled && !enableEC && length > chunkSize) {
```

Comment (guohao-rosicky) [reply hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java]: sure.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**